aptos/contract/aux/sources/reward_distributor.move

```diff
@@ -0,0 +1,325 @@
+module aux::reward_distributor {
+    use std::signer;
+
+    use aptos_framework::coin::{Self, Coin};
+    use aptos_framework::type_info;
+
+    const E_MISMATCH_DISTRIBUTOR_ID: u64 = 1002;
+    const E_CANNOT_CREATE_DISTRIBUTOR_FOR_OTHER: u64 = 1003;
+    const E_ZERO_MINT_AMOUNT: u64 = 1004;
+    const E_CANNOT_SHUTDOWN_DISTRIBUTOR_WITH_OUTSTANDING_TOKEN: u64 = 1005;
+    const E_REDEEM_TOKEN_NOT_FUNGIBLE: u64 = 1006;
+    const E_CANNOT_DESTROY_NON_ZERO_REDEEM_TOKEN: u64 = 1007;
+    const E_CANNOT_MELT_TOKEN_FOR_DIFFERNT_DISTRIBUTOR: u64 = 1008;
+
+    const DECIMAL_MULTIPLIER: u128 = 1000000000000;
+
+    /// RewardDistributor is a primitive for distributing rewards to token holders.
+    ///
+    /// Reward is given out in the form of 0x1::coin::Coin<R>.
+    ///
+    /// RewardDistrubor keeps track of acc_reward_per_share, that is,
+    /// the amount of reward per unit token if that token exists since the initialization of the distributor.
+    /// When reward is added to the RewardDistrubitor,
+    /// acc_reward_per_share is increased by the amount of added_reward / total_token_amount.
+    ///
+    ///
+    /// Users' rewards are represented by RedeemToken.
+    /// Users are entitled to rewards that are added after their RedeemToken are mint.
+    /// This is ensured by keeping in RedeemToken last_acc_reward_per_share, which is the
+    /// acc_reward_per_share when the RedeemToken is minted.
+    /// When RedeemTokens are burned, the returned reward will be based on
+    /// current RewardDistributor's acc_reward_per_share minus RedeemToken's last_acc_reward_per_share.
+    ///
+    /// In reality, acc_reward_per_share is multiplied by 10^12 to increase the precision of calculations.
+    /// However, rounding is unavoidable, and there may still be reward left when total token amount is 0.
+    ///
+    /// When total token amount drops to 0, RewardDistributor can be shutdown and rewards returned to the authority
+    /// of the distributor.
+    ///
+    /// W is the witness type. Only owner of the witness type can create a RewardDistributor with W.
+    ///
+    /// There can be multiple RewardDistributors for the same reward coin R and witness W. They are differentiated by their id.
+    /// Each RewardDistributor has separate authority even if they are of same reward coin R.
+    struct RewardDistributor<phantom R, phantom W> has store {
```

Comment:
> can we add an `authority: address` field to use for access control on sensitive functions?

`RewardDistributor` doesn't have a `key` capacity, so it must be stored inside some other object. so instead of guarding with an `address`, the functions can be guarded by the object itself. For example, `mint` requires `RewardDistributor`.

Reply:
Ah, right. So it's the responsibility of the calling contract to enforce proper access controls.